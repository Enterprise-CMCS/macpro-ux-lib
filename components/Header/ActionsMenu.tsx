import React, { useState } from "react";
import { Icon, IconChoice } from "../Icon/Icon";
import { Link, LinkProps } from "../Link/Link";

interface ActionsMenuLink extends LinkProps {
  iconName?: IconChoice;
}

interface Props {
  name: string;
  links: ActionsMenuLink[];
}

/**
 * ActionsMenu Component
 *
 * @param {string} name   The name of the menu.
 * @param {ActionsMenuLink[]} links  List of links to display in the menu.
 */

export const ActionsMenu: React.FC<Props> = ({ name, links, ...rest }) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div className="user-actions-container">
      <span className="menu-label" onClick={() => setIsOpen(!isOpen)}>
        {name}
        <Icon
          aria-hidden="true"
          name={isOpen ? "arrow_drop_up" : "arrow_drop_down"}
          color="#fff"
        />
      </span>
      <menu
        className="user-actions-menu usa-nav__secondary-links"
        style={isOpen ? { display: "block" } : {}}
        {...rest}
      >
        {links.map((linkData, idx) => {
          const { iconName, ...link } = linkData; // filter iconName out of linkData
          return (
            <li key={idx}>
              <Link {...link}>
                {iconName && <Icon name={iconName} />}
                {link.text}
              </Link>
            </li>
          );
        })}
      </menu>
    </div>
  );
};
