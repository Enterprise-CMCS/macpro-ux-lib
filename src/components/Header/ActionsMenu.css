.user-actions-drop-down li {
  cursor: pointer;
  margin-bottom: 10px;
}

.user-actions-drop-down li a {
  align-items: center;
  display: flex;
  flex-direction: row;
}
.user-actions-drop-down li a,
.menu-label {
  font-size: 1.06rem;
}

@media only screen and (min-width: 64em) {
  .user-actions-menu {
    color: #fff;
  }

  .user-actions-menu .menu-label {
    align-items: center;
    cursor: pointer;
    display: flex;
    flex-direction: row;
    font-weight: bold;
    position: relative;
  }

  .user-actions-drop-down {
    background-color: #0071bc;
    flex-direction: flex-start;
    left: 25px;
    position: absolute;
    top: 40px;
    width: 250px;
  }

  .user-actions-drop-down li {
    padding: 5px 15px 5px;
  }

  .user-actions-drop-down li a {
    color: #fff;
  }

  .user-actions-drop-down li a svg {
    color: #fff;
    margin-right: 10px;
  }

  .user-actions-drop-down li a:hover {
    color: #fff;
    text-decoration: unset;
  }

  .user-actions-drop-down {
    display: none;
    padding-top: 20px;
  }
}

@media only screen and (max-width: 64em) {
  .user-actions-menu {
    padding: 0;
  }
  .user-actions-menu a {
    padding: 12px 16px;
  }
  .user-actions-drop-down {
    display: block;
  }

  .user-actions-drop-down li {
    margin: 0;
    border-top: 1px solid #dfe1e2;
  }

  .user-actions-drop-down li a:hover,
  .user-actions-drop-down li a:hover svg {
    color: #005ea2;
    background-color: #f0f0f0;
    text-decoration: none;
  }

  .user-actions-drop-down li a svg {
    color: #565c65;
  }
}
