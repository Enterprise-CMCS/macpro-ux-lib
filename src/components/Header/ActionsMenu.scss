@use "../../assets/theme/macbis-colors" as c;

.user-actions-container {
  .user-actions-menu li {
    cursor: pointer;
  }

  .user-actions-menu a {
    align-items: center;
    display: flex;
    flex-direction: row;
  }
  .user-actions-menu a,
  .menu-label {
    font-size: 1.06rem;
  }

  .user-actions-menu svg {
    margin-right: 10px;
  }

  @media only screen and (min-width: 64em) {
    // USWDS Overrides
    .usa-nav__submenu,
    .usa-header--extended .usa-megamenu.usa-nav__submenu::after,
    .usa-header--extended .usa-megamenu.usa-nav__submenu::before {
      background-color: c.$primary-darkest;
    }

    .usa-nav__primary button {
      color: c.$gray;
    }

    .usa-header--extended .usa-nav__primary-item > .usa-current::after {
      background-color: c.$primary;
    }

    // New Component Styles
    .menu-label {
      color: #fff;
      align-items: center;
      cursor: pointer;
      display: flex;
      flex-direction: row;
      font-weight: bold;
      position: relative;
    }

    .user-actions-menu {
      display: none;
      background-color: c.$primary;
      flex-direction: flex-start;
      left: 25px;
      position: absolute;
      top: 41px;
      width: 250px;
    }

    .user-actions-menu a {
      padding: 10px 15px 10px;
    }

    .user-actions-menu {
      a,
      svg,
      a:hover {
        color: #fff;
        text-decoration: unset;
      }
    }

    .user-actions-menu a:hover,
    .user-actions-menu li:hover {
      background-color: c.$primary-darker;
    }
  }

  @media only screen and (max-width: 64em) {
    // USWDS Overrides
    .usa-menu-btn {
      background-color: c.$primary;
    }
    .usa-menu-btn:hover {
      background-color: c.$primary-darker;
    }

    .usa-nav__primary .usa-current,
    .usa-nav__primary .usa-current::after {
      color: c.$primary-darker;
    }

    .usa-nav__primary button {
      color: c.$gray;
    }

    .usa-nav__primary button:hover {
      color: c.$primary-darker;
      background-color: c.$gray-lightest;
    }

    .usa-nav__submenu-item {
      border-top: 1px solid c.$gray-lighter;
    }
    .usa-nav__primary a:not(.usa-button):not(.usa-current) {
      color: c.$gray;
    }

    .usa-nav__primary a:not(.usa-button):not(.usa-current):hover {
      color: c.$primary-darker;
    }

    // New Component Styles
    padding: 0;

    a {
      padding: 12px 16px;
    }

    .user-actions-menu {
      display: block;
    }

    .user-actions-menu,
    .user-actions-menu svg {
      color: c.$gray;
    }

    .user-actions-menu li {
      border-top: 1px solid c.$gray-lighter;
    }

    .user-actions-menu a:hover,
    .user-actions-menu a:hover svg {
      color: c.$primary-darker;
      background-color: c.$gray-lightest;
      text-decoration: none;
    }
  }
}
