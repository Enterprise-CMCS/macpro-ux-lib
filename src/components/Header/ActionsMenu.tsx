import React from "react";
import { Icon, IconChoice } from "components/Icon/Icon";

interface Link {
  href?: string;
  iconName?: IconChoice;
  onClick?: () => any;
  text: string;
}

interface Props {
  name: string;
  links: Link[];
}

/**
 * ActionsMenu Component
 *
 * @param {string} name   The name of the menu.
 * @param {Link[]} links  List of links to display in the menu.
 */

export const ActionsMenu: React.FC<Props> = ({ name, links, ...rest }) => {
  const [isOpen, setIsOpen] = React.useState(false);

  return (
    <div className="user-actions-container">
      <span className="menu-label" onClick={() => setIsOpen(!isOpen)}>
        {name}
        <Icon
          aria-hidden="true"
          name={isOpen ? "arrow_drop_up" : "arrow_drop_down"}
          color="#fff"
        />
      </span>
      <menu
        className="user-actions-menu usa-nav__secondary-links"
        style={isOpen ? { display: "block" } : {}}
        {...rest}
      >
        {links.map((link, idx) => {
          return (
            <li key={idx}>
              <a href={link.href} onClick={link.onClick}>
                {link.iconName && <Icon name={link.iconName} color="#fff" />}
                {link.text}
              </a>
            </li>
          );
        })}
      </menu>
    </div>
  );
};
