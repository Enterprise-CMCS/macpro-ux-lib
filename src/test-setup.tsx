import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
