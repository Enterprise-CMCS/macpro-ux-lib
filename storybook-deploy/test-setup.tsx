import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
