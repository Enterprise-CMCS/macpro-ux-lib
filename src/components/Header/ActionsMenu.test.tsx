import React from "react";
import { fireEvent, screen, render } from "../../test-setup";

import { ActionsMenu } from "./ActionsMenu";

const testOnClick = jest.fn();

const links = [
  {
    text: "Manage Profile",
    iconName: "person",
    href: "",
    onClick: testOnClick,
  },
  {
    text: "Request Role Change",
    iconName: "people",
    href: "",
    onClick: testOnClick,
  },
  {
    text: "Log Out",
    iconName: "logout",
    href: "",
    onClick: testOnClick,
  },
];

describe("Test the Header component", () => {
  it("should render", () => {
    render(<ActionsMenu name="My Account" links={links} />);

    // menu should contain the correct items
    ["Manage Profile", "Request Role Change", "Log Out"].forEach(
      (text, idx) => {
        const item = screen.getByText(text);
        expect(item).toHaveAttribute("href");
        fireEvent.click(item);
        expect(testOnClick).toHaveBeenCalledTimes(idx + 1);
      }
    );
  });

  it("menu should toggle open/close on click", () => {
    const { container } = render(
      <ActionsMenu name="My Account" links={links} />
    );
    // should render menu hidden

    // expand menu
    fireEvent.click(screen.getByText("My Account"));
  });

  it("menu should open on click", () => {
    const { container } = render(
      <ActionsMenu name="My Account" links={links} />
    );
    // should render menu hidden
    expect(container).toMatchSnapshot();

    // expand menu
    fireEvent.click(screen.getByText("My Account"));
    expect(container).toMatchSnapshot();
  });
});
