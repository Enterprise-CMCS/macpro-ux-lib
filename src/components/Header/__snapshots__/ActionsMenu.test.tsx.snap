// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the Header component menu should open on click 1`] = `
<div>
  <div
    class="user-actions-container"
  >
    <span
      class="menu-label"
    >
      My Account
      <svg
        aria-hidden="true"
        aria-label="arrow drop down icon"
        class="usa-icon--size-3"
        color="#fff"
        fill="currentColor"
        role="img"
      >
        <use
          href="sprite.svg#arrow_drop_down"
        />
      </svg>
    </span>
    <menu
      class="user-actions-menu usa-nav__secondary-links"
    >
      <li>
        <a
          href=""
        >
          <svg
            aria-hidden="true"
            aria-label="person icon"
            class="usa-icon--size-3"
            color="#fff"
            fill="currentColor"
            role="img"
          >
            <use
              href="sprite.svg#person"
            />
          </svg>
          Manage Profile
        </a>
      </li>
      <li>
        <a
          href=""
        >
          <svg
            aria-hidden="true"
            aria-label="people icon"
            class="usa-icon--size-3"
            color="#fff"
            fill="currentColor"
            role="img"
          >
            <use
              href="sprite.svg#people"
            />
          </svg>
          Request Role Change
        </a>
      </li>
      <li>
        <a
          href=""
        >
          <svg
            aria-hidden="true"
            aria-label="logout icon"
            class="usa-icon--size-3"
            color="#fff"
            fill="currentColor"
            role="img"
          >
            <use
              href="sprite.svg#logout"
            />
          </svg>
          Log Out
        </a>
      </li>
    </menu>
  </div>
</div>
`;

exports[`Test the Header component menu should open on click 2`] = `
<div>
  <div
    class="user-actions-container"
  >
    <span
      class="menu-label"
    >
      My Account
      <svg
        aria-hidden="true"
        aria-label="arrow drop up icon"
        class="usa-icon--size-3"
        color="#fff"
        fill="currentColor"
        role="img"
      >
        <use
          href="sprite.svg#arrow_drop_up"
        />
      </svg>
    </span>
    <menu
      class="user-actions-menu usa-nav__secondary-links"
      style="display: block;"
    >
      <li>
        <a
          href=""
        >
          <svg
            aria-hidden="true"
            aria-label="person icon"
            class="usa-icon--size-3"
            color="#fff"
            fill="currentColor"
            role="img"
          >
            <use
              href="sprite.svg#person"
            />
          </svg>
          Manage Profile
        </a>
      </li>
      <li>
        <a
          href=""
        >
          <svg
            aria-hidden="true"
            aria-label="people icon"
            class="usa-icon--size-3"
            color="#fff"
            fill="currentColor"
            role="img"
          >
            <use
              href="sprite.svg#people"
            />
          </svg>
          Request Role Change
        </a>
      </li>
      <li>
        <a
          href=""
        >
          <svg
            aria-hidden="true"
            aria-label="logout icon"
            class="usa-icon--size-3"
            color="#fff"
            fill="currentColor"
            role="img"
          >
            <use
              href="sprite.svg#logout"
            />
          </svg>
          Log Out
        </a>
      </li>
    </menu>
  </div>
</div>
`;
